dandi/dandifs.py

```diff
@@ -0,0 +1,434 @@
+    def client(self, x: DandiAPIClient):
+        if self.dandiset:
+            raise ValueError(
+                'Cannot assign a DANDI client to a FileSystem that is '
+                'already linked to a dandiset. Unassign the dandiset '
+                'first by: fs.dandiset = None')
+        self._client = x
+
+    @classmethod
+    def for_url(cls, url: str) -> "RemoteDandiFileSystem":
+        """
+        Instantiate a FileSystem that interacts with the correct
+        DANDI instance (and dandiset) for a given url.
+
+        :param url: absolute path, starting with with "https://",
+        "dandi://" or "DANDI:".
+        """
+        instance, dandiset, version, *_ = split_dandi_url(url)
+        return cls(dandiset, version, instance)
+
+    def get_dandiset(
+            self, path: str) -> Tuple[RemoteDandiset, Union[RemoteAsset, str]]:
+        """
+        If path is a relative path with respect to the current dandiset,
+        return (self.dandiset, path)
+
+        Else, the path is an absolute URL (it starts with "https://",
+        "dandi://" or "DANDI:"); we instantiate the correct remote
+        dandiset and spit out the relative path.
+
+        :param path: absolute or relative path to an asset
+        :returns: (dandiset, path) or (dandiset, asset)
+        """
+        dandiset = self.dandiset
+        if path.startswith(('http://', 'https://', 'dandi://', 'DANDI:')):
+            instance, dandiset_id, version_id, path, asset_id \
+                = split_dandi_url(path)
+            api_url = get_instance(instance)
+            if self.client.api_url == api_url.api:
+                client = self.client
+            else:
+                client = DandiAPIClient.for_dandi_instance(instance)
+                dandiset = None
+            if not asset_id:
+                if not dandiset or dandiset.identifier != dandiset_id:
+                    dandiset = client.get_dandiset(dandiset_id, version_id)
+                if not dandiset or dandiset.version_id != version_id:
+                    dandiset = client.get_dandiset(dandiset_id, version_id)
+            else:
+                asset = client.get_asset(asset_id)
+                return dandiset, asset
+        elif not self.dandiset:
+            raise ValueError('File system must be linked to a dandiset to '
+                             'use relative paths.')
+        return dandiset, path
+
+    def s3_url(self, path: str) -> str:
+        """
+        Get the the asset url on AWS S3
+
+        :param path: absolute or relative path to an asset
+        :returns: url to asset on S3
+        """
+        dandiset, asset = self.get_dandiset(path)
+        if not isinstance(asset, RemoteAsset):
+            asset = dandiset.get_asset_by_path(asset)
+        info = requests.request(url=asset.api_url, method='get').json()
+        url = ''
+        for url in info['contentUrl']:
+            if url.startswith('https://dandiarchive.s3.amazonaws.com'):
```

Comment:
Note that `dandiarchive.s3.amazonaws.com` is only used by the production `dandi` instance; the staging instance uses `dandi-api-staging-dandisets.s3.amazonaws.com`.